magento/magento2
# #23440 fix Refund for bundle product without receiving product back e…

### Description (*)

Fix Refund for bundle product without receiving the product back getting error


### Fixed Issues (if relevant)

1. Fixes magento/magento2#23440: Refund for bundle product without receiving product back

### Manual testing scenarios (*)

1. Create a bundle product with at least 2 products
2. Create order with bundle product
3. Invoice Order
4. Shop Order
5. Create Credit Memo without taking the product back in stock and only refund amount:
- Change all qty to 0
- Add adjustment refund: 10$
- Click on button update qty

![Screenshot - 2020-03-26T174254 117](https://user-images.githubusercontent.com/18417673/77638295-858c7980-6f89-11ea-8f1f-ca7ab0fae03e.png)

6. After click on update qty, bundle item not render and total refund amount is 25$
![Screenshot - 2020-03-26T174317 561](https://user-images.githubusercontent.com/18417673/77638292-84f3e300-6f89-11ea-8145-572f52bc6d13.png)

7. Create credit memo

Expected result (*)
Credit Memo for amount 25$ created

Actual result (*)

Error: "The most money available to refund is xx.xx.

![Screenshot - 2020-03-26T174352 267](https://user-images.githubusercontent.com/18417673/77638289-832a1f80-6f89-11ea-983a-4e60d0d1ac69.png)
### Addition info:




After check issue, I saw this issue happen because the bundle item does not render in the form when all child product qty set to 0

Check function "canRefundItem" in Magento/Sales/Model/Order/CreditmemoFactory.php:145 When we create a refund for bundle product with all item qty = 0

`
if ($item->getHasChildren()) {
                foreach ($item->getChildrenItems() as $child) {
                    if (empty($qtys)) {
                        if ($this->canRefundNoDummyItem($child, $invoiceQtysRefundLimits)) {
                            return true;
                        }
                    } else {
                        if (isset($qtys[$child->getId()]) && $qtys[$child->getId()] > 0) {
                            return true;
                        }
                    }
                }
               return false;
            }
`

The return value will return false and exclude bundle parent item in return result

=> bundle parent item does not render in Magento/Sales/view/adminhtml/templates/order/creditmemo/create/items.phtml
So the qty value for each item not exist
![Screenshot - 2020-03-26T174317 561](https://user-images.githubusercontent.com/18417673/77638292-84f3e300-6f89-11ea-8145-572f52bc6d13.png)

If the bundle item does not render in credit memo page and we create a credit memo, so when the function "createByOrder" to be call, the problem happen:
Magento/Sales/Model/Order/CreditmemoFactory.php:69

`        $qtyList = isset($data['qtys']) ? $data['qtys'] : [];`

 **$qtyList now is an empty array, but in this case, it should have data contain item_id with qty = 0**

 So this function will get all available refund qty for each item => the root of issue:
- If credit memo doesn't have adjustment amount, it will create credit memo for all item of bundle product (even we already updated all item qty to 0 - STEP 5 )

- If credit memo have adjustment amount, the error message will show: "The most money available to refund is xxxx"

I have pushed the commit to fix this issue
After apply update code, when we update qty for all bundle item to 0, the bundle item will show correctly:

![Screenshot - 2020-03-26T182313 566](https://user-images.githubusercontent.com/18417673/77641639-f1251580-6f8e-11ea-9da8-c9a0f72d8b96.png)

Result after create credit memo:

![Screenshot - 2020-03-26T182334 431](https://user-images.githubusercontent.com/18417673/77641665-fe420480-6f8e-11ea-941f-832c43e09f1d.png)


### Contribution checklist (*)
